src/lib/Inference.hs

```diff
@@ -809,27 +824,41 @@ getSrcCtx = lift ask
 
 -- === typeclass dictionary synthesizer ===
 
--- We have two variants here because at the top level we want error messages and
--- internally we want to consider all alternatives.
-type SynthPassM = SubstBuilderT (Either Err)
-type SynthDictM = SubstBuilderT []
-
-synthModule :: Scope -> Module -> Except Module
+synthModule :: Scope -> Module -> ExceptWithOutputs Module
 synthModule scope (Module Typed decls bindings) = do
   decls' <- fst . fst <$> runSubstBuilderT
-              (traverseDecls (traverseHoles synthDictTop) decls) scope
+              (traverseDecls (traverseHoles synthesize) decls) scope
   return $ Module Core decls' bindings
-synthModule _ _ = error $ "Unexpected IR variant"
-
-synthDictTop :: SrcCtx -> Type -> SynthPassM Atom
-synthDictTop ctx ty = do
+synthModule _ _ = error "Unexpected IR variant"
+
+-- Synthesis algorithm:
+-- 1. Run type inference for every possible instance, and filter out those
+--    that do not unify. At this point, do not recursively synthesize any
+--    dictionaries.
+-- 2. Check that we got exactly one instance.
```

Comment:
I don't think that this is an equivalently powerful algorithm to the old one. In particular, what if you had a hierarchy such as this one:
```
A a => B a
B a => C a
D a => E a
E a => C a
```
Now, consider a type that defines `E a`. In the first step of the search you cannot conclude failure just because you ended up with both `B a => C a` and `D a => E a` rules as applicable. Note that there is a single unique solution in this case and the previous algorithm would find it.

Reply:
Agreed, this is a less powerful algorithm. But I think that's the price you pay for better errors. Suppose a type `b` doesn't define anything, and then you request `C b`. What error do you show? If you're willing to do backtracking through everything, the best you can really do is say "I looked for `C b` but didn't find it". If you enforce that the synthesis algorithm only looks at one place in each iteration, you know exactly which missing instance was "responsible".

Haskell takes the non-backtracking approach, although I think some people have proposed various extensions that introduce backtracking in various principled ways.

This is related to #340, since in your example `B a => C a` and `E a => C a` can be seen as "overlapping" in the standard sense.

Reply:
I really hate the overlapping instances restriction in Haskell. I feel like it's been pushing a lot of people either towards `{-# LANGUAGE OverlappingInstances #-}` which is a terrible hack to make the algorithm proceed even when there are multiple matches, or some arcane type-level programming to work around this limitation. I'd rather avoid replicating the same issue here. What if we printed out all partial derivations? I don't expect that there will be quite enough matches anyway

Reply:
And even if there are more we can print something like:
```
Failed to derive the instance for <some expression>
The partial derivations are:
- <describe the derivation> (failed because an instance of <something> couldn't be found)
- <describe the derivation> (failed because an instance of <something> couldn't be found)
- <describe the derivation> (failed because an instance of <something> couldn't be found)
- ... (5 more derivations, specify `-some-flag` to print them all)
```

Reply:
Do you have real-world use cases similar to the scenario you mentioned above, where
- there are multiple possible strategies to derive `C a`,
- but only one of those strategies will succeed? (whereas overlapping instances in Haskell often is used to actually express a preference of one derivation over the other when multiple strategies succeed)

My personal feeling is that it's better for now to have clear error messages and an easily-interpretable algorithm, and then add more complexity back in once we actually have enough use cases. And my guess is that "unlimited backtracking but must find exactly one instance" isn't the right point to end up with, anyway; we probably want to handle overlapping instances even when multiple derivations succeed.

Reply:
After chatting with @dougalm, though, it seems like maybe it would be worthwhile to refactor the synthesis process a bit so that the algorithm/resolution strategy is orthogonal to the logging and inference machinery, since we may want to iterate on this in the future. I'll try to figure out something along those lines in the next week or so.

(Maybe a temporary compromise solution would be to run both the iterative and backtracking algorithms? If the iterative algorithm succeeds, use that; if it can't find anything, show a nice error; if it finds multiple things, then enable backtracking and see if that fixes the issue, but maybe show a worse error.)

Reply:
Looks like I was able to keep a fairly interpretable output while still doing backtracking. I'm still a bit skeptical that this is the "right" way to do it, but we can iterate on this more later if we have more concrete reasons to make a change.